File: backend/app/schemas/rbac.py
"""
Pydantic schemas for RBAC (Role-Based Access Control) system.

This module defines request/response schemas for the RBAC API endpoints
including validation rules and serialization formats.
"""

from pydantic import BaseModel, Field, validator
from typing import List, Optional, Set, Dict, Any
from datetime import datetime
from enum import Enum


class RoleHierarchyLevel(int, Enum):
    """Enumeration of role hierarchy levels."""
    VIEWER = 1
    USER = 2
    MANAGER = 3
    ADMIN = 4
    SUPER_ADMIN = 5


class PermissionAction(str, Enum):
    """Enumeration of permission actions."""
    READ = "read"
    CREATE = "create"
    UPDATE = "update"
    DELETE = "delete"
    ADMIN = "admin"


class ResourceType(str, Enum):
    """Enumeration of resource types."""
    DOCUMENTS = "documents"
    USERS = "users"
    ROLES = "roles"
    SYSTEM = "system"
    FOLDERS = "folders"


class SubjectType(str, Enum):
    """Enumeration of permission subject types."""
    USER = "user"
    ROLE = "role"


# Base schemas
class PermissionBase(BaseModel):
    """Base schema for permissions."""
    name: str = Field(..., min_length=3, max_length=100, description="Permission name (e.g., 'documents:read')")
    display_name: Optional[str] = Field(None, max_length=100, description="Human-readable permission name")
    description: Optional[str] = Field(None, description="Permission description")
    resource_type: str = Field(..., description="Resource type this permission applies to")
    action: str = Field(..., description="Action this permission allows")
    requires_resource_ownership: bool = Field(False, description="Whether permission requires owning the resource")

    @validator('name')
    def validate_permission_name(cls, v):
        """Validate permission name format."""
        if ':' not in v:
            raise ValueError("Permission name must follow format 'resource:action'")
        
        parts = v.split(':')
        if len(parts) != 2:
            raise ValueError("Permission name must follow format 'resource:action'")
        
        resource, action = parts
        if not resource or not action:
            raise ValueError("Both resource and action must be non-empty")
        
        return v.lower()


class PermissionCreate(PermissionBase):
    """Schema for creating new permissions."""
    pass


class PermissionUpdate(BaseModel):
    """Schema for updating permissions."""
    display_name: Optional[str] = Field(None, max_length=100)
    description: Optional[str] = None
    requires_resource_ownership: Optional[bool] = None


class Permission(PermissionBase):
    """Schema for permission responses."""
    id: int
    is_system: bool = Field(description="Whether this is a system permission")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Role schemas
class RoleBase(BaseModel):
    """Base schema for roles."""
    name: str = Field(..., min_length=2, max_length=50, description="Role name")
    display_name: Optional[str] = Field(None, max_length=100, description="Human-readable role name")
    description: Optional[str] = Field(None, description="Role description")
    hierarchy_level: Optional[int] = Field(None, ge=1, le=5, description="Role hierarchy level (1-5)")

    @validator('name')
    def validate_role_name(cls, v):
        """Validate role name format."""
        if not v.replace('_', '').isalnum():
            raise ValueError("Role name can only contain letters, numbers, and underscores")
        return v.lower()


class RoleCreate(RoleBase):
    """Schema for creating new roles."""
    permissions: Optional[List[str]] = Field(default_factory=list, description="List of permission names to assign")


class RoleUpdate(BaseModel):
    """Schema for updating roles."""
    display_name: Optional[str] = Field(None, max_length=100)
    description: Optional[str] = None
    permissions: Optional[List[str]] = Field(None, description="List of permission names to assign")


class Role(RoleBase):
    """Schema for role responses."""
    id: int
    hierarchy_level: int = Field(description="Role hierarchy level")
    is_system: bool = Field(description="Whether this is a system role")
    is_active: bool = Field(description="Whether this role is active")
    created_at: datetime
    updated_at: datetime
    permissions: List[Permission] = Field(default_factory=list, description="Permissions assigned to this role")

    class Config:
        from_attributes = True


class RoleWithStats(Role):
    """Schema for role responses with usage statistics."""
    user_count: int = Field(description="Number of users with this role")
    permission_count: int = Field(description="Number of permissions assigned to this role")


# User role assignment schemas
class UserRoleAssignmentBase(BaseModel):
    """Base schema for user role assignments."""
    user_id: int = Field(..., description="ID of the user")
    role_id: int = Field(..., description="ID of the role")
    is_primary: bool = Field(True, description="Whether this is the user's primary role")
    expires_at: Optional[datetime] = Field(None, description="When this role assignment expires")


class UserRoleAssignment(UserRoleAssignmentBase):
    """Schema for user role assignment requests."""
    pass


class UserRoleAssignmentResponse(UserRoleAssignmentBase):
    """Schema for user role assignment responses."""
    assigned_at: datetime
    assigned_by: Optional[int] = Field(None, description="ID of user who made the assignment")
    is_active: bool = Field(description="Whether this assignment is active")
    role: Role = Field(description="Role details")

    class Config:
        from_attributes = True


# Resource permission schemas
class ResourcePermissionBase(BaseModel):
    """Base schema for resource permissions."""
    resource_type: str = Field(..., description="Type of resource")
    resource_id: int = Field(..., description="ID of the specific resource")
    subject_type: SubjectType = Field(..., description="Type of subject (user or role)")
    subject_id: int = Field(..., description="ID of the subject")
    permission: str = Field(..., description="Permission level")
    granted: bool = Field(True, description="Whether permission is granted or denied")
    inheritable: bool = Field(True, description="Whether permission can be inherited")
    expires_at: Optional[datetime] = Field(None, description="When this permission expires")
    conditions: Optional[Dict[str, Any]] = Field(None, description="Conditional access rules")


class ResourcePermissionCreate(ResourcePermissionBase):
    """Schema for creating resource permissions."""
    pass


class ResourcePermissionUpdate(BaseModel):
    """Schema for updating resource permissions."""
    permission: Optional[str] = None
    granted: Optional[bool] = None
    inheritable: Optional[bool] = None
    expires_at: Optional[datetime] = None
    conditions: Optional[Dict[str, Any]] = None


class ResourcePermission(ResourcePermissionBase):
    """Schema for resource permission responses."""
    id: int
    granted_at: datetime
    granted_by: Optional[int] = Field(None, description="ID of user who granted permission")
    inherited_from: Optional[int] = Field(None, description="ID of parent permission if inherited")

    class Config:
        from_attributes = True


# Permission check schemas
class PermissionCheckRequest(BaseModel):
    """Schema for permission check requests."""
    user_id: int = Field(..., description="ID of the user")
    permission: str = Field(..., description="Permission to check")
    resource_type: Optional[str] = Field(None, description="Type of resource")
    resource_id: Optional[int] = Field(None, description="ID of specific resource")
    context: Optional[Dict[str, Any]] = Field(None, description="Additional context for permission check")


class PermissionCheckResponse(BaseModel):
    """Schema for permission check responses."""
    granted: bool = Field(description="Whether permission is granted")
    reason: str = Field(description="Reason for grant or denial")
    source: str = Field(description="Source of permission (role, resource, etc.)")
    expires_at: Optional[datetime] = Field(None, description="When permission expires")


# Bulk operation schemas
class BulkRoleAssignment(BaseModel):
    """Schema for bulk role assignments."""
    user_ids: List[int] = Field(..., min_items=1, description="List of user IDs")
    role_name: str = Field(..., description="Role name to assign")
    is_primary: bool = Field(False, description="Whether to set as primary role")
    expires_at: Optional[datetime] = Field(None, description="When assignments expire")


class BulkRoleAssignmentResponse(BaseModel):
    """Schema for bulk role assignment responses."""
    successful: List[int] = Field(description="User IDs successfully assigned")
    failed: List[Dict[str, Any]] = Field(description="Failed assignments with reasons")
    total_processed: int = Field(description="Total number of users processed")


class BulkPermissionGrant(BaseModel):
    """Schema for bulk permission grants."""
    subject_type: SubjectType = Field(..., description="Type of subjects")
    subject_ids: List[int] = Field(..., min_items=1, description="List of subject IDs")
    resource_type: str = Field(..., description="Type of resources")
    resource_ids: List[int] = Field(..., min_items=1, description="List of resource IDs")
    permission: str = Field(..., description="Permission to grant")
    expires_at: Optional[datetime] = Field(None, description="When permissions expire")


# Audit and reporting schemas
class RoleAuditEntry(BaseModel):
    """Schema for role audit log entries."""
    id: int
    user_id: int
    role_name: str
    action: str = Field(description="Action performed (assign, revoke, create, delete)")
    performed_by: int = Field(description="ID of user who performed the action")
    performed_at: datetime
    details: Optional[Dict[str, Any]] = Field(None, description="Additional details")

    class Config:
        from_attributes = True


class PermissionAuditEntry(BaseModel):
    """Schema for permission audit log entries."""
    id: int
    user_id: int
    permission: str
    resource_type: Optional[str] = None
    resource_id: Optional[int] = None
    granted: bool = Field(description="Whether permission was granted")
    checked_at: datetime
    source: str = Field(description="Source of permission decision")

    class Config:
        from_attributes = True


class AccessDeniedLog(BaseModel):
    """Schema for access denied log entries."""
    id: int
    user_id: int
    attempted_action: str
    resource_type: Optional[str] = None
    resource_id: Optional[int] = None
    reason: str = Field(description="Reason for denial")
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    attempted_at: datetime

    class Config:
        from_attributes = True


# Statistics and reporting schemas
class RoleStatistics(BaseModel):
    """Schema for role usage statistics."""
    role_name: str
    user_count: int = Field(description="Number of users with this role")
    permission_count: int = Field(description="Number of permissions assigned")
    hierarchy_level: int
    is_active: bool
    created_at: datetime


class UserPermissionSummary(BaseModel):
    """Schema for user permission summary."""
    user_id: int
    username: str
    primary_role: Optional[str] = Field(None, description="Primary role name")
    all_roles: List[str] = Field(description="All assigned role names")
    permission_count: int = Field(description="Total number of permissions")
    highest_hierarchy_level: int = Field(description="Highest role hierarchy level")
    last_login: Optional[datetime] = None


class SystemPermissionMatrix(BaseModel):
    """Schema for system-wide permission matrix."""
    roles: List[Role] = Field(description="All system roles")
    permissions: List[Permission] = Field(description="All system permissions")
    matrix: Dict[str, List[str]] = Field(description="Role -> Permission mappings")
    hierarchy: Dict[str, int] = Field(description="Role hierarchy levels")


# Response wrapper schemas
class RoleListResponse(BaseModel):
    """Schema for paginated role list responses."""
    roles: List[RoleWithStats]
    total: int = Field(description="Total number of roles")
    page: int = Field(description="Current page number")
    size: int = Field(description="Page size")
    has_next: bool = Field(description="Whether there are more pages")


class PermissionListResponse(BaseModel):
    """Schema for paginated permission list responses."""
    permissions: List[Permission]
    total: int = Field(description="Total number of permissions")
    page: int = Field(description="Current page number")
    size: int = Field(description="Page size")
    has_next: bool = Field(description="Whether there are more pages")


class UserRoleListResponse(BaseModel):
    """Schema for user role list responses."""
    user_roles: List[UserRoleAssignmentResponse]
    total: int = Field(description="Total number of role assignments")
    user_id: int = Field(description="User ID")
    username: str = Field(description="Username")


# Error schemas
class RBACError(BaseModel):
    """Schema for RBAC-specific errors."""
    error_code: str = Field(description="Error code")
    message: str = Field(description="Error message")
    details: Optional[Dict[str, Any]] = Field(None, description="Additional error details")


class ValidationError(BaseModel):
    """Schema for validation errors."""
    field: str = Field(description="Field that failed validation")
    message: str = Field(description="Validation error message")
    rejected_value: Any = Field(description="Value that was rejected")


# Configuration schemas
class RBACConfig(BaseModel):
    """Schema for RBAC system configuration."""
    enable_resource_permissions: bool = Field(True, description="Enable resource-level permissions")
    enable_temporary_roles: bool = Field(True, description="Enable temporary role assignments")
    enable_conditional_permissions: bool = Field(False, description="Enable conditional permissions")
    default_role: str = Field("user", description="Default role for new users")
    audit_all_checks: bool = Field(False, description="Audit all permission checks")
    cache_permissions: bool = Field(True, description="Cache permission lookups")
    cache_ttl_seconds: int = Field(300, description="Permission cache TTL in seconds")

File: frontend/src/types/rbac.ts
/**
 * RBAC (Role-Based Access Control) Type Definitions
 * 
 * This file contains TypeScript interfaces and types for the RBAC system,
 * including roles, permissions, user assignments, and API responses.
 */

export interface Role {
  id: number;
  name: string;
  display_name: string;
  description: string;
  hierarchy_level: number;
  is_system: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  created_by?: number;
  permissions?: Permission[];
}

export interface RoleWithStats extends Role {
  user_count: number;
  permission_count: number;
}

export interface Permission {
  id: number;
  name: string;
  display_name: string;
  description: string;
  resource_type: string;
  action: string;
  is_system: boolean;
  requires_resource_ownership: boolean;
  created_at: string;
  updated_at: string;
}

export interface UserRole {
  user_id: number;
  role_id: number;
  assigned_at: string;
  assigned_by?: number;
  expires_at?: string;
  is_primary: boolean;
  is_active: boolean;
  role: Role;
}

export interface ResourcePermission {
  id: number;
  resource_type: string;
  resource_id: number;
  subject_type: 'user' | 'role';
  subject_id: number;
  permission: string;
  granted: boolean;
  granted_at: string;
  granted_by: number;
  expires_at?: string;
  conditions?: Record<string, any>;
}

// API Request/Response Types
export interface RoleCreate {
  name: string;
  display_name: string;
  description: string;
  hierarchy_level?: number;
  permissions?: string[];
}

export interface RoleUpdate {
  display_name?: string;
  description?: string;
  permissions?: string[];
}

export interface RoleListResponse {
  roles: Role[] | RoleWithStats[];
  total: number;
  page: number;
  size: number;
  has_next: boolean;
}

export interface PermissionCreate {
  name: string;
  display_name: string;
  description: string;
  resource_type: string;
  action: string;
  requires_resource_ownership?: boolean;
}

export interface PermissionUpdate {
  display_name?: string;
  description?: string;
  requires_resource_ownership?: boolean;
}

export interface PermissionListResponse {
  permissions: Permission[];
  total: number;
  page: number;
  size: number;
  has_next: boolean;
}

export interface UserRoleAssignment {
  user_id: number;
  role_id: number;
  is_primary?: boolean;
  expires_at?: string;
}

export interface UserRoleAssignmentResponse extends UserRole {
  // Includes all UserRole fields
}

export interface UserRoleListResponse {
  user_roles: UserRole[];
  total: number;
  user_id: number;
  username: string;
}

export interface BulkRoleAssignment {
  user_ids: number[];
  role_name: string;
  is_primary?: boolean;
  expires_at?: string;
}

export interface BulkRoleAssignmentResponse {
  successful: number[];
  failed: Array<{
    user_id: number;
    error: string;
  }>;
  total_processed: number;
}

export interface PermissionCheckRequest {
  user_id: number;
  permission: string;
  resource_type?: string;
  resource_id?: number;
}

export interface PermissionCheckResponse {
  granted: boolean;
  reason: string;
  source: 'role' | 'resource' | 'none';
}

export interface UserPermissionSummary {
  user_id: number;
  username: string;
  primary_role?: string;
  all_roles: string[];
  permission_count: number;
  highest_hierarchy_level: number;
  last_login?: string;
}

export interface SystemPermissionMatrix {
  roles: Role[];
  permissions: Permission[];
  matrix: Record<string, string[]>;
  hierarchy: Record<string, number>;
}

export interface ResourcePermissionCreate {
  resource_type: string;
  resource_id: number;
  subject_type: 'user' | 'role';
  subject_id: number;
  permission: string;
  granted: boolean;
  expires_at?: string;
  conditions?: Record<string, any>;
}

export interface ResourcePermissionUpdate {
  granted?: boolean;
  expires_at?: string;
  conditions?: Record<string, any>;
}

export interface RBACError {
  detail: string;
  error_code?: string;
  field_errors?: Record<string, string[]>;
}

// Role hierarchy constants
export const ROLE_HIERARCHY = {
  VIEWER: 1,
  USER: 2,
  MANAGER: 3,
  ADMIN: 4,
  SUPER_ADMIN: 5
} as const;

export const ROLE_NAMES = {
  [ROLE_HIERARCHY.VIEWER]: 'Viewer',
  [ROLE_HIERARCHY.USER]: 'User', 
  [ROLE_HIERARCHY.MANAGER]: 'Manager',
  [ROLE_HIERARCHY.ADMIN]: 'Admin',
  [ROLE_HIERARCHY.SUPER_ADMIN]: 'Super Admin'
} as const;

// Permission action constants
export const PERMISSION_ACTIONS = {
  READ: 'read',
  CREATE: 'create',
  UPDATE: 'update',
  DELETE: 'delete',
  ADMIN: 'admin'
} as const;

// Resource type constants
export const RESOURCE_TYPES = {
  DOCUMENTS: 'documents',
  USERS: 'users',
  ROLES: 'roles',
  SYSTEM: 'system',
  FOLDERS: 'folders'
} as const;

// UI State types
export interface RoleManagementState {
  roles: Role[];
  selectedRole?: Role;
  isLoading: boolean;
  error?: string;
  showCreateDialog: boolean;
  showEditDialog: boolean;
  showDeleteDialog: boolean;
  filters: {
    search: string;
    hierarchy_level?: number;
    active_only: boolean;
  };
  pagination: {
    page: number;
    size: number;
    total: number;
    has_next: boolean;
  };
}

export interface UserRoleAssignmentState {
  users: Array<{
    id: number;
    username: string;
    email: string;
    account_locked: boolean;
    last_login: string | null;
    roles: UserRole[];
  }>;
  selectedUser?: number;
  availableRoles: Role[];
  isLoading: boolean;
  error?: string;
  showAssignDialog: boolean;
  showBulkDialog: boolean;
  bulkSelection: number[];
}

export interface PermissionMatrixState {
  roles: Role[];
  permissions: Permission[];
  matrix: Record<string, string[]>;
  hierarchy: Record<string, number>;
  isLoading: boolean;
  error?: string;
  filters: {
    resource_type?: string;
    role_hierarchy?: number;
  };
}

// Form validation types
export interface RoleFormData {
  name: string;
  display_name: string;
  description: string;
  hierarchy_level: number;
  permissions: string[];
}

export interface RoleFormErrors {
  name?: string;
  display_name?: string;
  description?: string;
  hierarchy_level?: string;
  permissions?: string;
}

// Component props types
export interface RoleManagementProps {
  showStats?: boolean;
  allowCreate?: boolean;
  allowEdit?: boolean;
  allowDelete?: boolean;
  onRoleSelect?: (role: Role) => void;
}

export interface UserRoleAssignmentProps {
  userId?: number;
  showBulkOperations?: boolean;
  onAssignmentChange?: () => void;
}

export interface PermissionMatrixProps {
  interactive?: boolean;
  showHierarchy?: boolean;
  highlightRole?: string;
  onPermissionClick?: (role: string, permission: string) => void;
  compact?: boolean;
  mobileView?: 'matrix' | 'summary' | 'hierarchy';
}

export interface AccessDeniedProps {
  requiredPermission?: string;
  requiredRole?: string;
  message?: string;
  showReturnButton?: boolean;
  onReturnClick?: () => void;
}

// Utility types
export type RoleHierarchyLevel = typeof ROLE_HIERARCHY[keyof typeof ROLE_HIERARCHY];
export type PermissionAction = typeof PERMISSION_ACTIONS[keyof typeof PERMISSION_ACTIONS];
export type ResourceType = typeof RESOURCE_TYPES[keyof typeof RESOURCE_TYPES];
export type SubjectType = 'user' | 'role';

// API service types
export interface RBACServiceInterface {
  // Role management
  getRoles(params?: {
    page?: number;
    size?: number;
    include_stats?: boolean;
    active_only?: boolean;
  }): Promise<RoleListResponse>;
  
  getRole(roleId: number): Promise<Role>;
  createRole(data: RoleCreate): Promise<Role>;
  updateRole(roleId: number, data: RoleUpdate): Promise<Role>;
  deleteRole(roleId: number): Promise<void>;
  
  // Permission management
  getPermissions(params?: {
    page?: number;
    size?: number;
    resource_type?: string;
  }): Promise<PermissionListResponse>;
  
  createPermission(data: PermissionCreate): Promise<Permission>;
  
  // User role assignments
  getUserRoles(userId: number): Promise<UserRoleListResponse>;
  assignRoleToUser(userId: number, assignment: UserRoleAssignment): Promise<UserRoleAssignmentResponse>;
  revokeRoleFromUser(userId: number, roleId: number): Promise<void>;
  bulkAssignRoles(assignment: BulkRoleAssignment): Promise<BulkRoleAssignmentResponse>;
  
  // Permission checking
  checkPermission(request: PermissionCheckRequest): Promise<PermissionCheckResponse>;
  getUserPermissions(userId: number): Promise<string[]>;
  getUserPermissionSummary(userId: number): Promise<UserPermissionSummary>;
  
  // System utilities
  getPermissionMatrix(): Promise<SystemPermissionMatrix>;
  
  // Resource permissions
  createResourcePermission(data: ResourcePermissionCreate): Promise<ResourcePermission>;
  getResourcePermissions(resourceType: string, resourceId: number): Promise<ResourcePermission[]>;
  
  // Health check
  checkHealth(): Promise<{ status: string; roles: number; permissions: number; timestamp: string }>;
}